### PR TITLE
[CALCITE-3542] Implement RepeatUnion All=false

### PR DESCRIPTION
**Pull request by rubenada**
Jira ticket: [CALCITE-3542](https://issues.apache.org/jira/browse/CALCITE-3542)  

Previously, [CALCITE-2812](https://issues.apache.org/jira/browse/CALCITE-2812) introduced recursive queries in Calcite via the new operator RepeatUnion.
However, this initial implementation was not complete, since it only provided RepeatUnion ALL (i.e. duplicates were returned).
The current PR aims to implement the missing feature: RepeatUnion non-ALL (i.e. duplicates must be discarded).

### CODE REVIEW

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java [hunk 1]
```diff
@@ -3845,67 +3850,94 @@ public void close() {
           private final Enumerator<TSource> seedEnumerator = seed.enumerator();
           private Enumerator<TSource> iterativeEnumerator = null;
 
+          // Set to control duplicates, only used if all is false
```
Comment (jinxing64) [inline linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Nit: shall we give a double quotes for `all` ?

**Comment by jinxing64** [comment]
LGTM

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 1 above)
Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Agree, change committed.
Thanks for the review @jinxing64 !

File: core/src/test/java/org/apache/calcite/test/enumerable/EnumerableRepeatUnionHierarchyTest.java [hunk 2]
```diff
@@ -55,33 +58,60 @@
...
   }
 
-  @ParameterizedTest
+  @ParameterizedTest(name = "{index} : hierarchy(startIds:{2}, ascendant:{3}, "
+      + "maxDepth:{4}, all:{0})")
   @MethodSource("data")
-  public void testHierarchy(int startId, boolean ascendant,
-                            int maxDepth, String[] expected) {
+  public void testHierarchy(boolean all, int[] startIds, String startIdsStr,
+                            boolean ascendant, int maxDepth, String[] expected) {
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/test/enumerable/EnumerableRepeatUnionHierarchyTest.java]: The indentation looks weird.

Comment (rubenada) [reply core/src/test/java/org/apache/calcite/test/enumerable/EnumerableRepeatUnionHierarchyTest.java]: Agree, fixed.

**Pull request merged**